File: dynamic_programming/0494_target_sum/0494_target_sum.go
package main

import (
	"fmt"
)

func findTargetSumWays(nums []int, S int) int {
	sum := 0

	for i := 0; i < len(nums); i++ {
		sum += nums[i]
	}

	if sum < S {
		return 0
	}

	if (sum+S)%2 == 1 {
		return 0
	}

	return calcSumWays(nums, (sum+S)/2)

}

func calcSumWays(nums []int, target int) int {
	dp := make([]int, target+1)
	dp[0] = 1
	for i := 0; i < len(nums); i++ {
		for j := target; j >= nums[i]; j-- {
			dp[j] += dp[j-nums[i]]
		}
	}
	return dp[target]
}

func main() {
	nums := []int{1, 1, 1, 1, 1}
	target := 3
	res := findTargetSumWays(nums, target)
	fmt.Println(res)
}

File: dynamic_programming/0494_target_sum/0494_target_sum.py

class Solution(object):
    def findTargetSumWays(self, nums, S):
        from collections import defaultdict
        memo = {0:1}
        for x in nums:
            m = defaultdict(int)
            for s, n in memo.iteritems():
                m[s + x] += n
                m[s - x] += n
            memo = m
            print memo
        return memo[S]


nums = [1,1,1,1,1]
S = 3
res = Solution().findTargetSumWays(nums, S)
print(res)